### PR TITLE
add support for filters on sg and subnet

### PR DESCRIPTION
**Pull request by enxebre**
This allows to integrate the openshift actuator with mao and the installer:
- https://github.com/enxebre/installer/blob/mao/modules/bootkube/resources/manifests/mao-config.yaml#L3
- https://github.com/openshift/machine-api-operator/pull/34

### CODE REVIEW

**Comment by enxebre** [comment]
Follow up:
- Add test case for machine-with-filters.yaml as we fix https://github.com/openshift/cluster-api-provider-aws/issues/29 cc @spangenberg

File: examples/machine-with-filters.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      publicIp: true
+      iamInstanceProfile:
+        id: meh-master-profile
+      keyName: tectonic
+      tags:
+        - name: "kubernetes.io/cluster/meh"
+          value: owned
+      securityGroups:
+        - filters:
+          - name: "tag:Name"
```
Comment (ingvagabund) [inline examples/machine-with-filters.yaml]: Are filters fixed to tags or can you use non-tag names to specify filters?

File: cloud/aws/actuators/machine/actuator.go [hunk 2]
```diff
@@ -104,6 +104,73 @@ func (a *Actuator) removeStoppedMachine(machine *clusterv1.Machine, client awscl
...
+	for _, g := range securityGroups {
+		// ID has priority
+		if g.ID != nil {
+			securityGroupIDs = append(securityGroupIDs, g.ID)
+		} else if g.Filters != nil {
+			mLog.Debug("Describing security groups based on filters")
+			ec2Filters := buildEc2Filters(g.Filters)
+			// Get groups based on filters
+			describeSecurityGroupsRequest := ec2.DescribeSecurityGroupsInput{
+				Filters: ec2Filters,
```
Comment (ingvagabund) [inline cloud/aws/actuators/machine/actuator.go]: `ec2Filters` is not used anywhere else so it's fine to `Filters: buildEc2Filters(g.Filters)`

File: cloud/aws/actuators/machine/actuator.go [hunk 3]
```diff
@@ -104,6 +104,73 @@ func (a *Actuator) removeStoppedMachine(machine *clusterv1.Machine, client awscl
...
+func getSubnetID(subnet providerconfigv1.AWSResourceReference, client awsclient.Client, mLog log.FieldLogger) (*string, error) {
+	var subnetIDs []*string
+	// ID has priority
+	if subnet.ID != nil {
+		subnetIDs = append(subnetIDs, subnet.ID)
+	} else {
+		mLog.Debug("Describing subnets based on filters")
+		ec2Filters := buildEc2Filters(subnet.Filters)
+		describeSubnetRequest := ec2.DescribeSubnetsInput{
+			Filters: ec2Filters,
```
Comment (ingvagabund) [inline cloud/aws/actuators/machine/actuator.go]: `ec2Filters` is not used anywhere else so it's fine to `Filters: buildEc2Filters(subnet.Filters)`

File: cloud/aws/actuators/machine/actuator.go [hunk 4]
```diff
@@ -210,7 +280,7 @@ func (a *Actuator) CreateMachine(cluster *clusterv1.Cluster, machine *clusterv1.
 	}
 	tagList = append(tagList, []*ec2.Tag{
 		{Key: aws.String("clusterid"), Value: aws.String(cluster.Name)},
-		{Key: aws.String("kubernetes.io/cluster/" + cluster.Name), Value: aws.String(cluster.Name)},
+		{Key: aws.String("kubernetes.io/cluster/" + cluster.Name), Value: aws.String("owned")},
```
Comment (ingvagabund) [inline cloud/aws/actuators/machine/actuator.go]: Why `owned`?

File: cloud/aws/actuators/machine/actuator.go (hunk 3 above)
Comment (ingvagabund) [reply cloud/aws/actuators/machine/actuator.go]: `buildEc2Filters` looks good, can you use it in the AMI selection part as well? Lines 143-157.

File: cloud/aws/actuators/machine/actuator.go (hunk 4 above)
Comment (enxebre) [reply cloud/aws/actuators/machine/actuator.go]: this matches current installer behaviour

File: examples/machine-with-filters.yaml (hunk 1 above)
Comment (enxebre) [reply examples/machine-with-filters.yaml]: you can specify aws attributes, not necessarily tags, hence we don't wanna harcode it in code

File: cloud/aws/actuators/machine/actuator.go [hunk 5]
```diff
@@ -104,6 +104,71 @@ func (a *Actuator) removeStoppedMachine(machine *clusterv1.Machine, client awscl
...
+			Filters: buildEc2Filters(subnet.Filters),
+		}
+		describeSubnetResult, err := client.DescribeSubnets(&describeSubnetRequest)
+		if err != nil {
+			mLog.Errorf("error describing security groups: %v", err)
+			return nil, fmt.Errorf("error describing security groups: %v", err)
+		}
+		for _, n := range describeSubnetResult.Subnets {
+			subnetID := *n.SubnetId
+			subnetIDs = append(subnetIDs, &subnetID)
```
Comment (dgoodwin) [inline cloud/aws/actuators/machine/actuator.go]: Looks like you could just return here on the first one and stop iterating the loop if we're just looking to return the first.

Should we warn if multiple results were returned?

Comment (enxebre) [reply cloud/aws/actuators/machine/actuator.go]: thanks! made the function able to return multiple ids, then upper level function decideds what to do and warns

**Pull request merged**